src/server/system_services/master_key_manager.js

```diff
@@ -0,0 +1,246 @@
+/* Copyright (C) 2016 NooBaa */
+'use strict';
+
+const _ = require('lodash');
+const util = require('util');
+const crypto = require('crypto');
+const config = require('../../../config');
+const db_client = require('../../util/db_client').instance();
+const dbg = require('../../util/debug_module')(__filename);
+const SensitiveString = require('../../util/sensitive_string');
+const LRUCache = require('../../util/lru_cache');
+
+const ROOT_KEY = 'noobaa-root-master-key';
+
+class MasterKeysManager {
+    constructor() {
+        this.is_initialized = false;
+        this.resolved_master_keys_by_id = {};
+        this.master_keys_by_id = {};
+        this.secret_keys_cache = new LRUCache({
+            name: 'SecretKeysCache',
+            expiry_ms: Infinity,
+            max_usage: 5000,
```

Comment:
In theory, I wouldn't want to limit, we can use Infinity.
Shouldn't have plenty of keys anyway.